### ISSUE TITLE
Building custom dataloaders

### ISSUE THREAD

**Issue by oussemajelassi** [MaximIntegratedAI/ai8x-training#200]
I am looking into building a custom dataloader, 
I am following this documentation : [Data Loader Design for MAX78000 Model Training](https://www.analog.com/en/app-notes/data-loader-design-for-max78000-model-training.html#ref-2)

coming to this part : 'Expected Data Range
For training, input data is expected to be in the range  . When evaluating quantized weights, or when running on hardware, input data is instead expected to be in the native MAX7800X range of [-128, +127].

As described in the following sections, the data loader function takes the data path and some arguments as input arguments. The arguments field includes two required fields, act_mode_8bit and **truncate_testset**. When set to True, the first argument refers to the case normalization should be done correctly for the native MAX7800X range, i.e., to range [-128, +127]. When set to False, the normalization should be in the range of  for training. ' 

1/ I was not able to understand the role of truncate_testset. 

2/ Dataloaders will take part in training phase and eval phase how can i change that argument for each phase.

3/ After building my own dataloader what is the next step in order to make my own model trained and deployed on MAX78000

Thanks in advance, A LOT !

**Comment by seldauyanik-maxim** [comment]
Dear Oussama

Here are my comments:

1) truncate_testset is used to get only the first image from the test set:

   When set, the test_dataset data loader returns will only have a single item. This can be handy for testing model output for single item in a resource-constrained environment and for debugging purposes.
   Noting that the AISegment implementation is revised for clarifying that part out further by PR 203
   
2) train.py handles automatic argument settings for each phase. But if you have a custom training code/notebook you should set them accordingly.
    
   Example data loader creation for training using `distiller.apputils`:
```
    simulate = False
    data_path = '[some_path here]'
    args = Args(act_mode_8bit=simulate)
    batch_size = 16
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    train_loader, val_loader, test_loader, _ =
        apputils.get_data_loaders(aisegment.AISegment80_get_datasets,
                                  (data_path, args), batch_size=16, cpu=device == 'cpu')
```
   For the evaluation mode, you need to set simulate parameter to `True`. Also, `get_data_loaders` has some additional parameters you can take a look into and set according to your needs.
   
3) For the next steps in order to make your own model trained and deployed on MAX78000, I would suggest going through the related Readme parts: 
      - General Overview: https://github.com/MaximIntegratedAI/ai8x-training#overview
      - Keep in mind the limitations for hardware validation while designing your model: https://github.com/MaximIntegratedAI/ai8x-training#limitations-of-max78000-networks
      - Adding new models: https://github.com/MaximIntegratedAI/ai8x-training#adding-new-network-models-and-new-datasets-to-the-training-process
      - Training section: https://github.com/MaximIntegratedAI/ai8x-training#limitations-of-max78000-networks
      - Usage of network loader a.k.a. izer: https://github.com/MaximIntegratedAI/ai8x-training#network-loader-ai8xize

   Briefly can be summarized as follows:
   - I advise testing data loader implementation in a separate custom test code first to retrieve some samples from the data loader and check some other functionalities as well along with the expected data range, types, etc.
   - Design your model (see examples in models/ folder)
   - Add training script: Take a look at the training scripts that reside under scripts/train_*
	   `--dataset` parameter is specifying your dataset
	   `--model` specifies your model
	   Also, there can be some additional files you may want to modify like qat_policy, LR scheduling, etc. and add related options to train.py 
   - Once your model is trained, you can perform some manual model output checks and visualizations.	   
   - Use the synthesis repository to quantize your trained model and auto generate a sample C code for the known answer test that you can deploy on hardware.

**Comment by oussemajelassi** [comment]
Thank you.

**Comment by oussemajelassi** [comment]
Hello again In Your examples in dataset folder I see no actual creating of a dataloader, do train.py handles that in an automatic way ? 
I mean there is no call for `train_dataloader = DataLoader(training_data, batch_size=64, shuffle=True)` for example.

**Comment by seldauyanik-maxim** [comment]
Train.py creates data loader using --dataset argument, for details see
- lines ~400 in train py and
- get_data_loaders function in distiller/apputils/data_loaders.py
